File: arena/python/arena.py
"""Python interface to CPG Arena (standard library only).

    from arena import Problem

    p = Problem("creatures/worm.toml", mode="race")
    p.dim                          # number of genes; a genome is p.dim floats in [0, 1]
    p.start                        # the template creature as a genome
    p.evaluate(population)         # list of genomes -> list of fitness (higher is better)
    p.save(best, "me.toml", name="My Creature")

Each `evaluate` call runs the whole population in parallel in one process,
so call it once per generation rather than once per individual.
"""

import json
import os
import shutil
import subprocess
from pathlib import Path

__all__ = ["Problem"]


def _find_binary():
    if os.environ.get("ARENA_BIN"):
        return os.environ["ARENA_BIN"]
    here = Path(__file__).resolve().parent.parent
    for name in ("arena", "arena.exe"):
        p = here / "target" / "release" / name
        if p.exists():
            return str(p)
    found = shutil.which("arena")
    if found:
        return found
    raise FileNotFoundError("arena binary not found: run `cargo build --release` in arena/ or set ARENA_BIN")


class Problem:
    """A creature-evolution problem.

    template:  creature file whose body plan is kept; genes fill in the numbers
    mode:      "race" (distance in metres) or "sumo" (win/loss + ring control)
    body:      also let genes change segment sizes, attach points and angles
    opponents: sumo only, folder of creatures to fight (default: the Rock)
    rules:     rules file (default: arena.toml next to the template)
    """

    def __init__(self, template, mode="race", body=False, opponents=None, rules=None):
        self._bin = _find_binary()
        self._args = [str(template), "--mode", mode]
        if body:
            self._args.append("--body")
        if opponents:
            self._args += ["--opponents", str(opponents)]
        if rules:
            self._args += ["--rules", str(rules)]
        self.info = json.loads(self._run("info", "--json"))
        self.dim = self.info["dim"]
        self.genes = [g["name"] for g in self.info["genes"]]
        self.start = [g["start"] for g in self.info["genes"]]
        self.evaluations = 0

    def _run(self, cmd, *extra, stdin=None):
        res = subprocess.run([self._bin, cmd, *self._args, *extra], input=stdin, capture_output=True, text=True)
        if res.returncode != 0:
            raise ValueError(res.stderr.strip())
        return res.stdout

    def evaluate(self, population):
        """Fitness of every genome in `population` (a list of lists of floats in [0, 1])."""
        for k, g in enumerate(population):
            if len(g) != self.dim:
                raise ValueError(f"genome {k} has {len(g)} genes, expected {self.dim}")
        text = "\n".join(",".join(repr(float(x)) for x in g) for g in population)
        out = self._run("batch", stdin=text)
        self.evaluations += len(population)
        return [float(line) for line in out.split()]

    def save(self, genome, path, name=None):
        """Write the creature for `genome` to `path`; returns its fitness."""
        extra = ["--genes", ",".join(repr(float(x)) for x in genome), "--out", str(path)]
        if name:
            extra += ["--name", name]
        out = self._run("save", *extra)
        return float(out.split()[1])

    def describe(self, genome):
        """Human-readable gene values, handy for debugging."""
        lines = []
        for g, x in zip(self.info["genes"], genome):
            lines.append(f"{g['name']:<22} {g['lo'] + x * (g['hi'] - g['lo']):8.2f}")
        return "\n".join(lines)

File: arena/src/bin/arena.rs
//! Command-line tool. The optimisation interface for your own GA:
//!
//! ```text
//! arena info  worm.toml                 how many genes, what they mean
//! arena batch worm.toml < pop.txt       one genome per line in, one fitness per line out
//! arena eval  worm.toml --genes 0.1,…   fitness of one genome
//! arena save  worm.toml --genes 0.1,… --out me.toml --name "Me"
//! ```
//!
//! Plus `check`, `race` and `tournament` for folders of creatures.

use clap::{Args, Parser, Subcommand};
use cpg_arena::creature::Rules;
use cpg_arena::game::{self, Mode};
use cpg_arena::problem::Problem;
use std::io::{BufRead, BufWriter, Write};
use std::path::{Path, PathBuf};

#[derive(Parser)]
#[command(name = "arena", about = "CPG Arena: evaluate genomes for your own genetic algorithm, race and fight creatures")]
struct Cli {
    #[command(subcommand)]
    cmd: Cmd,
}

/// What is being optimised. Same flags for info / eval / batch / save.
#[derive(Args)]
struct ProblemArgs {
    /// Template creature: its body plan (topology) is fixed, the genes fill in the numbers.
    template: PathBuf,
    #[arg(long, value_enum, default_value = "race")]
    mode: Mode,
    /// Genes also control the body (segment sizes, attach points, angles).
    #[arg(long)]
    body: bool,
    /// Sumo only: fight every creature in this folder (default: the Rock).
    #[arg(long)]
    opponents: Option<PathBuf>,
    /// Rules file (default: arena.toml next to the template, else built-in rules).
    #[arg(long)]
    rules: Option<PathBuf>,
}

impl ProblemArgs {
    fn problem(&self) -> Problem {
        Problem::load(&self.template, self.mode, self.body, self.opponents.as_deref(), self.rules.as_deref()).unwrap_or_else(|e| die(&e))
    }
}

#[derive(Subcommand)]
enum Cmd {
    /// Describe the genome: number of genes, names, ranges, template values.
    Info {
        #[command(flatten)]
        problem: ProblemArgs,
        /// Machine-readable output.
        #[arg(long)]
        json: bool,
    },
    /// Fitness of one genome (higher is better).
    Eval {
        #[command(flatten)]
        problem: ProblemArgs,
        /// Genes in [0,1], comma-separated. Default: the template itself.
        #[arg(long, value_delimiter = ',', allow_negative_numbers = true)]
        genes: Option<Vec<f64>>,
    },
    /// Fitness of a whole population, evaluated in parallel.
    /// Reads one genome per line from stdin (numbers separated by commas or
    /// spaces), writes one fitness per line to stdout, in the same order.
    Batch {
        #[command(flatten)]
        problem: ProblemArgs,
    },
    /// Turn a genome into a creature file for the arena.
    Save {
        #[command(flatten)]
        problem: ProblemArgs,
        #[arg(long, value_delimiter = ',', allow_negative_numbers = true)]
        genes: Vec<f64>,
        #[arg(long)]
        out: PathBuf,
        /// Name shown in the arena (default: the template's name).
        #[arg(long)]
        name: Option<String>,
    },
    /// Validate every creature in a folder.
    Check { folder: PathBuf },
    /// Race every creature in a folder.
    Race { folder: PathBuf },
    /// Sumo round robin between every creature in a folder.
    Tournament { folder: PathBuf },
}

fn die(msg: &str) -> ! {
    eprintln!("error: {msg}");
    std::process::exit(1)
}

fn folder_rules(dir: &Path) -> Rules {
    Rules::load_dir(dir).unwrap_or_else(|e| die(&e))
}

fn parse_genome(line: &str) -> Result<Vec<f64>, String> {
    line.split(|c: char| c == ',' || c.is_whitespace())
        .filter(|t| !t.is_empty())
        .map(|t| t.parse::<f64>().map_err(|e| format!("'{t}': {e}")))
        .collect()
}

fn main() {
    let cli = Cli::parse();
    match cli.cmd {
        Cmd::Info { problem, json } => {
            let p = problem.problem();
            let info = p.info();
            if json {
                println!("{}", serde_json::to_string_pretty(&info).expect("serialisable"));
                return;
            }
            println!("creature  {}\nmode      {}\nbody      {}\ndim       {}", info.creature, info.mode, info.body, info.dim);
            if p.mode == Mode::Sumo {
                let ops = if info.opponents.is_empty() { "Rock".to_string() } else { info.opponents.join(", ") };
                println!("opponents {ops}");
            }
            println!("\n  #  {:<22} {:>16}  {:>6}", "gene", "0 → 1 maps to", "start");
            for (i, g) in info.genes.iter().enumerate() {
                println!("{i:>3}  {:<22} [{:>6.2}, {:>6.2}]  {:>6.3}", g.name, g.lo, g.hi, g.start);
            }
        }
        Cmd::Eval { problem, genes } => {
            let p = problem.problem();
            let g = genes.unwrap_or_else(|| p.start());
            println!("{}", p.evaluate(&g).unwrap_or_else(|e| die(&e)));
        }
        Cmd::Batch { problem } => {
            let p = problem.problem();
            let mut pop = Vec::new();
            for (n, line) in std::io::stdin().lock().lines().enumerate() {
                let line = line.unwrap_or_else(|e| die(&e.to_string()));
                if line.trim().is_empty() || line.trim_start().starts_with('#') {
                    continue;
                }
                pop.push(parse_genome(&line).unwrap_or_else(|e| die(&format!("line {}: {e}", n + 1))));
            }
            let fit = p.evaluate_batch(&pop).unwrap_or_else(|e| die(&e));
            let mut out = BufWriter::new(std::io::stdout().lock());
            for f in fit {
                writeln!(out, "{f}").ok();
            }
        }
        Cmd::Save { problem, genes, out, name } => {
            let p = problem.problem();
            let f = p.save(&genes, &out, name.as_deref()).unwrap_or_else(|e| die(&e));
            println!("fitness {f} -> {}", out.display());
        }
        Cmd::Check { folder } => {
            let rules = folder_rules(&folder);
            let mut bad = 0;
            for e in game::load_folder(&folder, &rules) {
                match &e.creature {
                    Ok(c) => println!("ok    {:<24} {} segments, area {:.3} m²", c.name, c.segments.len(), c.area()),
                    Err(errs) => {
                        bad += 1;
                        println!("FAIL  {}", e.path.display());
                        for m in errs {
                            println!("        {m}");
                        }
                    }
                }
            }
            if bad > 0 {
                std::process::exit(1);
            }
        }
        Cmd::Race { folder } => {
            let rules = folder_rules(&folder);
            let cs: Vec<_> = game::load_folder(&folder, &rules).into_iter().filter_map(|e| e.creature.ok()).collect();
            for (rank, (i, d)) in game::race_all(&cs, &rules).into_iter().enumerate() {
                println!("{:>2}. {:<24} {:>7.2} m", rank + 1, cs[i].name, d);
            }
        }
        Cmd::Tournament { folder } => {
            let rules = folder_rules(&folder);
            let cs: Vec<_> = game::load_folder(&folder, &rules).into_iter().filter_map(|e| e.creature.ok()).collect();
            let (table, _) = game::tournament(&cs, &rules);
            println!("    {:<24} {:>3} {:>3} {:>3} {:>4}", "", "W", "D", "L", "Pts");
            for (rank, s) in table.iter().enumerate() {
                println!("{:>2}. {:<24} {:>3} {:>3} {:>3} {:>4}", rank + 1, cs[s.index].name, s.wins, s.draws, s.losses, s.points);
            }
        }
    }
}
